### ISSUE TITLE
Large buffer size leads to packet drops

### ISSUE THREAD

**Issue by sla89** [the-tcpdump-group/libpcap#843]
Hello together,

I am capturing packets using libpcap 1.9.0 on linux (kernel 4.19.50-rt22) and I am facing the problem that I have packet drops when I am using buffer sizes >=4MiB.The larger the value the more packets were dropped. The snapshot length was set to 1538 bytes.

With a buffer size below the mentioned value I have no packet drops at all. 

Can you please explain to me why larger buffer sizes lead to drops? In my understanding a larger buffer size should avoid packet drops.

I have the same behavior when I am running tcpdump with the mentioned values on that platform.

It would be great to get an explanation for this behavior.

Thank you very much and with best regards,
Stefan

**Comment by Oppen** [comment]
Maybe buffers this big thrash the TLBs? Just a guess. If it's this, using huge pages should help, but I'm not sure if libpcap provides that optino.
Could you test a dummy read-and-discard on both scenarios?
I think that could give us a lead on whether libpcap is doing something wrong or the application code starts performing worse (thus queuing more packets) on bigger buffers.

**Comment by sla89** [comment]
What do you mean with dummy read-and-discard?

**Comment by Oppen** [comment]
Doing something like:
```
void dummy_handler(u_char *user, const struct pcap_pkthdr *h, const u_char *bytes) {
  return;
}
int main(void) {
  // Initialization stuff...
pcap_loop(pcap_inst, -1, dummy_handler, NULL);
  return 0;
```

**Comment by Oppen** [comment]
Essentially, do no processing, to check whether your processing became slower or libpcap is misbehaving.

**Comment by guyharris** [comment]
> Maybe buffers this big thrash the TLBs? Just a guess. If it's this, using huge pages should help, but I'm not sure if libpcap provides that optino.

It doesn't, and I'm not sure it should; instead, it should, if possible, figure out by itself whether huge pages can be used and, if so, request the appropriate size.

Assuming that the kernel will allow the packet buffer of a PF_PACKET socket to be mapped with huge pages, is there any reason *not* to just have libpcap decide whether to use them?

**Comment by Oppen** [comment]
> It doesn't, and I'm not sure it should; instead, it should, if possible, figure out by itself whether huge pages can be used and, if so, request the appropriate size.
Sounds reasonable.
> Assuming that the kernel will allow the packet buffer of a PF_PACKET socket to be mapped with huge pages, is there any reason _not_ to just have libpcap decide whether to use them?
Sounds doable.
Should we assume only 2MiB pages are allowed? 1GiB pages are reserved at boot time, so I don't think a user would want us to take them without permission.

**Comment by sla89** [comment]
> Doing something like:
> 
> ```
> void dummy_handler(u_char *user, const struct pcap_pkthdr *h, const u_char *bytes) {
>   return;
> }
> int main(void) {
>   // Initialization stuff...
> pcap_loop(pcap_inst, -1, dummy_handler, NULL);
>   return 0;
> ```

I already did this when I was doing my tests.

I am using `pcap_dispatch` with `-1` to receive all packets at once.

Before calling `pcap_dispatch` I set up my environment (in the following order):

- set snapshot length to 1538bytes
- set buffer size to 16MB
- enabled promiscuous mode
- set buffer timeout to 10ms
- enable non-blocking mode

**Comment by Oppen** [comment]
Great, those details will be useful.
If you're willing to build from source to compare, I can probably attach a little patch with a proof of concept fix, against 1.9.0 for reproducibility.
It will require `/sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages` to be >= 2048 (preferably something a bit greater) and `hugetlbfs` to be mounted somewhere.
This should do:
```
#!/bin/sh

echo 2560 | sudo tee /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages > /dev/null
sudo mkdir /hugetlb
sudo mount -t hugetlbfs hugetlbfs /hugetlb
```

**Comment by Oppen** [comment]
Well, here goes nothing. I have to mention, I only made sure it builds.
I inline it because GitHub is curiously not friendly towards attaching patches.
```
From 4577d17cb3617d0c64541016d33a8cfd4b1f7699 Mon Sep 17 00:00:00 2001
From: Mario Rugiero <mario.rugiero@helpsystems.com>
Date: Tue, 1 Oct 2019 17:29:01 -0300
Subject: [PATCH] PoC huge pages on Linux.

---
 pcap-linux.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/pcap-linux.c b/pcap-linux.c
index 49144b64..9bd62e55 100644
--- a/pcap-linux.c
+++ b/pcap-linux.c
@@ -4620,7 +4620,7 @@ retry:
        /* memory map the rx ring */
        handlep->mmapbuflen = req.tp_block_nr * req.tp_block_size;
        handlep->mmapbuf = mmap(0, handlep->mmapbuflen,
-           PROT_READ|PROT_WRITE, MAP_SHARED, handle->fd, 0);
+           PROT_READ|PROT_WRITE, MAP_SHARED | MAP_HUGETLB, handle->fd, 0);
        if (handlep->mmapbuf == MAP_FAILED) {
                pcap_fmt_errmsg_for_errno(handle->errbuf, PCAP_ERRBUF_SIZE,
                    errno, "can't mmap rx ring");
@@ -4674,7 +4674,8 @@ destroy_ring(pcap_t *handle)
        /* if ring is mapped, unmap it*/
        if (handlep->mmapbuf) {
                /* do not test for mmap failure, as we can't recover from any error */
-               (void)munmap(handlep->mmapbuf, handlep->mmapbuflen);
+               size_t mmaplen_2m = ((handlep->mmapbuflen + (1 << 21) - 1) >> 21) << 21;
+               (void)munmap(handlep->mmapbuf, mmaplen_2m);
                handlep->mmapbuf = NULL;
        }
 }
-- 
2.20.1
```

**Comment by Oppen** [comment]
OK, I *think* I got something. Are you using `pcap_set_buffer_size` to set the buffer?
```
int
pcap_set_buffer_size(pcap_t *p, int buffer_size)
{
        if (pcap_check_activated(p))
                return (PCAP_ERROR_ACTIVATED);
        if (buffer_size <= 0) {
                /*
                 * Silently ignore invalid values.
                 */
                return (0);
        }
        p->opt.buffer_size = buffer_size;
        return (0);
}
```
See how it uses a signed `int` and then ignores non-positive values? 4096MiB are the same as `1 << 32` is the same bit pattern as `0` for `int`.
Thus, it uses the default buffer size, which is 2MiB. Your application drops because the buffer fills up rather quickly.
This is a limitation in the API, and according to #651, it isn't fixed, nor planned to be fixed.
This will also be a problem with DPDK, where it will probably be a common need to use big buffers.

Nonetheless, using huge pages may be interesting.

**Comment by guyharris** [comment]
> This is a limitation in the API, and according to #651, it isn't fixed, nor planned to be fixed.

The current buffer size setting routine can't be fixed, as it takes an int as an argument.

A new API would have to be added, perhaps taking a `size_t` as an argument.  The activate routines on various platforms for various devices would have to deal appropriately with a buffer size too big to be handled by the OS or device APIs used to set it (e.g., if the size doesn't fit in a `u_int` on systems with BPF, where the argument to the `BIOCSBLEN` ioctl is a `u_int`).

> This will also be a problem with DPDK, where it will probably be a common need to use big buffers.

Then perhaps it's time to add that new API.

**Comment by Oppen** [comment]
> Then perhaps it's time to add that new API.

I believe that would require some degree of discussion. What would be the proper place for it?

**Comment by guyharris** [comment]
> > Then perhaps it's time to add that new API.
> 
> I believe that would require some degree of discussion. What would be the proper place for it?

Either 1) this issue or 2) tcpdump-workers@lists.tcpdump.org.

The latter would require that all participants be members of that list, but it'd also open it up to people unaware of this particular GitHub issue.

**Comment by Oppen** [comment]
I guess we can expect readers of this comment interested in participating to subscribe to the list?
I'll be doing that right now, and create a thread soon (either in a few hours or tomorrow).
Worst case, we could post a link to the thread in the mail archives, right?

**Comment by Oppen** [comment]
Regarding the original topic: the best you can get is passing `INT_MAX` to `pcap_set_buffer_size`.

**Comment by guyharris** [comment]
> I guess we can expect readers of this comment interested in participating to subscribe to the list?

They should; [see here for information on how to subscribe](https://www.tcpdump.org/#mailing-lists).  It's not a high-volume list (unlike, say, the torrent that is linux-netdev; I wish I could subscribe to linux-netdev-but-just-the-PF_PACKET-and-non-driver-specific-link-layer-stuff, as that's the only stuff that matters to me as a libpcap developer).

> I'll be doing that right now, and create a thread soon (either in a few hours or tomorrow).
> Worst case, we could post a link to the thread in the mail archives, right?

I.e., post it here for the benefit of those not subscribed?  Yes, we could.

**Comment by Oppen** [comment]
> I wish I could subscribe to linux-netdev-but-just-the-PF_PACKET-and-non-driver-specific-link-layer-stuff, as that's the only stuff that matters to me as a libpcap developer).

We can always use filters on our clients. I never do, though.

> I.e., post it here for the benefit of those not subscribed? Yes, we could.

Exactly.

**Comment by sla89** [comment]
Thank you very much for the discussion and your suggestions!

Yes, I am using `set_buffer_size` to set the buffer size but I think the current limitation of a maximum buffer size of 2GiB is not tackling me because I am just setting it to 16MiB, right? 
What my tests showed me was that larger buffer sizes (like 32MiB or 64MiB) will lead to more drops when the same traffic is received. I never tried to set it to 2GiB or more. 64MiB was the maximum. 

Sorry that I have missed this details at the beginning.

@Oppen thanks for sharing your patch with me. It will take me a few days to have it fully tested. I will definitely come back to you.

**Comment by Oppen** [comment]
Oh. The original post mentioned 4096M, was that a typo, or did you later try the smaller sizes to help diagnosing?
Can you tell me the device and driver you're using?
Ideas for more tests:
Try to reproduce it with virtual interfaces. I'm AWK, but tell me if you need help setting that up.
Can you provide a pcap that triggers the issue?

**Comment by sla89** [comment]
Sorry, this was definitely a typo. I was talking about 4MiB. I have fixed it in the initial post.

I am using an Intel i350 with the igb driver.

I can easily reproduce it using the pktgen module by running its sample:

```
./pktgen_sample03_burst_single_flow.sh -i ens5f1 -m FF:FF:FF:FF:FF:FF -d 192.168.1.1 -n 1000000 -s 1500
```

**Comment by guyharris** [comment]
> Or is it valid to just add a new error code?

Yes, it is.

EINVAL is *not* one of my favorite UN*X errors.  The more specific the error reported by an error code, the better.

**Comment by Oppen** [comment]
> EINVAL is _not_ one of my favorite UN*X errors. The more specific the error reported by an error code, the better.

Does `PCAP_ERROR_BUFFER_TOO_BIG` sound good?

**Comment by guyharris** [comment]
> Does `PCAP_ERROR_BUFFER_TOO_BIG` sound good?

Yes.

**Comment by Oppen** [comment]
Sent the RFC. Can follow the thread [here](https://lists.sandelman.ca/pipermail/tcpdump-workers/2019-October/001290.html).

**Comment by Oppen** [comment]
I was able to reproduce the issue with veth interfaces.
I used `pktgen_sample01_simple.sh` instead due to lack of support of hw bursts on such devices.
```
Buffer size:    2B bytes - packet loss: 7.973618%
Buffer size:    4B bytes - packet loss: 7.754280%
Buffer size:    8B bytes - packet loss: 8.034293%
Buffer size:   16B bytes - packet loss: 8.002143%
Buffer size:   32B bytes - packet loss: 7.813963%
Buffer size:   64B bytes - packet loss: 7.969014%
Buffer size:  128B bytes - packet loss: 8.946679%
Buffer size:  256B bytes - packet loss: 8.292493%
Buffer size:  512B bytes - packet loss: 7.956970%
Buffer size:    1k bytes - packet loss: 11.465474%
Buffer size:    2k bytes - packet loss: 7.550734%
Buffer size:    4k bytes - packet loss: 8.446791%
Buffer size:    8k bytes - packet loss: 8.400310%
Buffer size:   16k bytes - packet loss: 7.253298%
Buffer size:   32k bytes - packet loss: 8.263069%
Buffer size:   64k bytes - packet loss: 7.657072%
Buffer size:  128k bytes - packet loss: 8.047637%
Buffer size:  256k bytes - packet loss: 8.357261%
Buffer size:  512k bytes - packet loss: 0.001878%
Buffer size:    1M bytes - packet loss: 0.000545%
Buffer size:    2M bytes - packet loss: 0.004727%
Buffer size:    4M bytes - packet loss: 0.014421%
Buffer size:    8M bytes - packet loss: 0.035376%
Buffer size:   16M bytes - packet loss: 0.079915%
Buffer size:   32M bytes - packet loss: 0.169770%
Buffer size:   64M bytes - packet loss: 0.328689%
Buffer size:  128M bytes - packet loss: 0.669605%
Buffer size:  256M bytes - packet loss: 1.338953%
Buffer size:  512M bytes - packet loss: 2.637323%
Buffer size: 1024M bytes - packet loss: 5.112182%
```
For some reason, loss seems linearly proportional to buffer size.
I ran a modified `capturetest` that made 10000 iterations of `pcap_disaptch` and then printed the stats shown above.
Loss is calculated as `(double)(100 * ps.ps_drop)/ps.ps_recv`.

**Comment by Oppen** [comment]
After many hours of experimenting, I have good news and bad news.
The good news is that I can reproduce the issue, as shown below.
The bad news is that it seems like PACKET_MMAP does its mapping outside of the scope where we can tell it how the mapping should be, so we can't ask for it to be composed of huge pages.
I haven't figured out the issue yet.
Here are my results:
```
Buffer size:    2B bytes - packet loss: 7.973618%
Buffer size:    4B bytes - packet loss: 7.754280%
Buffer size:    8B bytes - packet loss: 8.034293%
Buffer size:   16B bytes - packet loss: 8.002143%
Buffer size:   32B bytes - packet loss: 7.813963%
Buffer size:   64B bytes - packet loss: 7.969014%
Buffer size:  128B bytes - packet loss: 8.946679%
Buffer size:  256B bytes - packet loss: 8.292493%
Buffer size:  512B bytes - packet loss: 7.956970%
Buffer size:    1k bytes - packet loss: 11.465474%
Buffer size:    2k bytes - packet loss: 7.550734%
Buffer size:    4k bytes - packet loss: 8.446791%
Buffer size:    8k bytes - packet loss: 8.400310%
Buffer size:   16k bytes - packet loss: 7.253298%
Buffer size:   32k bytes - packet loss: 8.263069%
Buffer size:   64k bytes - packet loss: 7.657072%
Buffer size:  128k bytes - packet loss: 8.047637%
Buffer size:  256k bytes - packet loss: 8.357261%
Buffer size:  512k bytes - packet loss: 0.001878%
Buffer size:    1M bytes - packet loss: 0.000545%
Buffer size:    2M bytes - packet loss: 0.004727%
Buffer size:    4M bytes - packet loss: 0.014421%
Buffer size:    8M bytes - packet loss: 0.035376%
Buffer size:   16M bytes - packet loss: 0.079915%
Buffer size:   32M bytes - packet loss: 0.169770%
Buffer size:   64M bytes - packet loss: 0.328689%
Buffer size:  128M bytes - packet loss: 0.669605%
Buffer size:  256M bytes - packet loss: 1.338953%
Buffer size:  512M bytes - packet loss: 2.637323%
Buffer size: 1024M bytes - packet loss: 5.112182%
```
Observations:
- Buffer size indicates the parameter, not necessarily the actual size. In practice, minimum buffer size is 256kiB, that's why loss vary very little before that.
- Loss seems to linearly increase with the size of the buffer afterwards.
- The big drop from 256kiB to 512kiB is most likely attributed to the fact it is now able to store more than one frame at a time.
- The way it's programmed, snaplen is basically ignored, using always the maximum size.

**Comment by guyharris** [comment]
> * The big drop from 256kiB to 512kiB is most likely attributed to the fact it is now able to store more than one frame at a time.
> * The way it's programmed, snaplen is basically ignored, using always the maximum size.
> * Any attempts at using a bigger block size for the ring buffer lead to dramatically degraded performance (+1000x in terms of captured packets).

So TPACKET_V3 is more complicated than BPF ("BPF" in the "capture mechanism" sense, not in the "interpreted pseudo-machine code supported by that capture mechanism and some other capture mechanisms" sense).

With BPF, you get two buffers, each of which is the size specified by the `SIOCSBLEN` ioctl.  When one buffer fills up - or is marked "finished" by a timeout - it's made available to read, and the other buffer, if emptied by the userland code, is set up to be the one into which packets are placed.  If the  buffer being filled fills up or is marked finished when the other buffer hasn't been emptied, any packets that arrive are dropped.

With TPACKET_V3, you have N blocks, each of a spacified size M.  Currently, we make all blocks `MAXIMUM_SNAPLEN` in size, and request a number of buffers equal to the buffer size, rounded up to a multiple of `MAXIMUM_SNAPLEN`, and divided by `MAXIMUM_SNAPLEN`.

`MAXIMUM_SNAPLEN` is currently 256 KiB; that size was chosen for the benefit of Linux USB captures (captures of USB traffic, not captures on an interface that happens to be connected via USB), so it's probably bigger than need be for most regular interfaces.  It's probably big enough for at least 128 full-size Ethernet packet, for example (I'm too lazy to calculate the per-packet overhead of TPACKET_V3 and add that in :-)).

So, yeah, with a block size of 256 KiB, that's one buffer, which userland had better drain *really* quickly in order not to get drops.

By "block size" in "bigger block size" are you referring to the size of a block, i.e. 256 KiB in the current code?  And by "bigger" do you mean "just doubling it to 512 KiB makes matters worse, even with a 1 MiB buffer"?

For BPF, macOS has a minimum buffer size of 32 (!) bytes.  That's probably not going to work well unless you have a *really* small snapshot length - a minimum-sized Ethernet packet will get sliced off (24 of those 32 bytes would go to the BPF header, so you wouldn't even get a full *Ethernet* header!).

For Linux, perhaps the block size should be set such that you get at least two blocks, even if that means you can get the same kind of silliness you get from a minimum-sized BPF buffer size.

**Comment by sla89** [comment]
@Oppen can you please share your script with me so that I can run it on my real hardware so that we can compare the results?

Thanks for the investigation and your time!

**Comment by Oppen** [comment]
> `MAXIMUM_SNAPLEN` is currently 256 KiB; that size was chosen for the benefit of Linux USB captures (captures of USB traffic, not captures on an interface that happens to be connected via USB), so it's probably bigger than need be for most regular interfaces. It's probably big enough for at least 128 full-size Ethernet packet, for example (I'm too lazy to calculate the per-packet overhead of TPACKET_V3 and add that in :-)).
> 
> So, yeah, with a block size of 256 KiB, that's one buffer, which userland had better drain _really_ quickly in order not to get drops.
> 

That I noticed, but I didn't know the reason. It should be painless to make the program aware whether it is USB capture and setting a more reasonable snaplen otherwise.
I'll try reducing it to 64kiB just for experimentation.

> By "block size" in "bigger block size" are you referring to the size of a block, i.e. 256 KiB in the current code? And by "bigger" do you mean "just doubling it to 512 KiB makes matters worse, even with a 1 MiB buffer"?
> 

Block size as in tp_block_size, as requested to the socket.
I did try with the same loop that tried buffer sizes of all powers of 2 from 2 to 1Gi, and the block is specified in the docs to contain as many full frames as they fit which, since we're working with powers of 2 and 256kiB is a power of 2, won't change the total size, because twice the block size means twice the full frames in each block and the fact the buffer size remains constant means we have half as many blocks.
TL;DR, I still fail to see what causes the performance drop.

> For Linux, perhaps the block size should be set such that you get at least two blocks, even if that means you can get the same kind of silliness you get from a minimum-sized BPF buffer size.

Yes, that makes a lot of sense.


> @Oppen can you please share your script with me so that I can run it on my real hardware so that we can compare the results?
> 

Give me one or two hours, I made adjustments here and there on the road and need to make it more consistent.
Also, many changes required directly modifying the library, as settings aren't exposed to users.

> Thanks for the investigation and your time!

Glad to be of help :)

**Comment by Oppen** [comment]
For injection I used:
```
./pktgen_sample01_simple.sh -i veth0 -m FF:FF:FF:FF:FF:FF -d 192.168.1.1 -n 0 -s 1500
```
I'd say you use the same for reproducibility. Burst didn't work for me, I believe because of not using real hardware.

I used a modified `capturetest` for capture, inlined.
Note I removed the copyright notice for brevity, but it is still a derivative, and thus apply the same license conditions.
I'd say you just replace 'capturetest.c' on you repo with this, so `make` will make life easier for you.
```
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include <pcap.h>

static void
countme(u_char *user, const struct pcap_pkthdr *h, const u_char *sp)
{
        int *counterp = (int *)user;

        (*counterp)++;
}

static pcap_t *pd;

int
main(int argc, char **argv)
{
        int op;
        char *device;
        int immediate = 0;
        int snaplen = 1536;
        int buffer_size = (1 << 18);
        const int timeout = 1000;
        int status;
        int packet_count;
        int packet_count_ac = 0;

        opterr = 0;
        while ((op = getopt(argc, argv, "b:i:ms:")) != -1) {
                switch (op) {
                case 'b':
                        buffer_size = atoi(optarg);
                        break;
                case 'i':
                        device = optarg;
                        break;
                case 'm':
                        immediate = 1;
                        break;
                case 's':
                        snaplen = atoi(optarg);
                        break;
                }
        }

        if (device == NULL)
                return 1;

        pd = pcap_create(device, NULL);
        if (pd == NULL)
                return 2;
        status = pcap_set_snaplen(pd, snaplen);
        if (status != 0)
                return 3;
        if (immediate) {
                status = pcap_set_immediate_mode(pd, 1);
                if (status != 0)
                        return 4;
        }
        pcap_set_buffer_size(pd, buffer_size); /* Never fails. */
        status = pcap_set_timeout(pd, timeout);
        if (status != 0)
                return 5;

        status = pcap_activate(pd);
        if (status < 0)
                return 6;

        for (size_t i = 0; i < 10000; ++i) {
                packet_count = 0;
                status = pcap_dispatch(pd, -1, countme,
                    (u_char *)&packet_count);
                if (status < 0)
                        break;
                if (status == 0)
                        continue;
                packet_count_ac += packet_count;
        }
        struct pcap_stat ps;
        pcap_stats(pd, &ps);
        char buffer_unit = 'B';
        if (buffer_size >> 10) {
                buffer_size >>= 10;
                buffer_unit = 'k';
        }
        if (buffer_size >> 10) {
                buffer_size >>= 10;
                buffer_unit = 'M';
        }
        fprintf(stderr, "buffer size: %4d%c bytes - packet loss: %3.6f%% total captured packets: %d\n",
            buffer_size, buffer_unit, (double)(100 * ps.ps_drop)/ps.ps_recv, packet_count_ac);
        pcap_close(pd);
        return 0;
}
```
Lastly, I used the following script to test powers of 2 as buffer sizes:
```
#!/bin/bash

for i in {1..60}; do echo >&2; done

for i in {1..30}; do
        ./capturetest -b $((2**$i)) -i veth1
        echo >&2
done
```

**Comment by sla89** [comment]
Thanks, really appreciate it! I will come back with my results as soon as possible.

**Comment by sla89** [comment]
I am sorry for the very late response. I was very busy because we did a release of our software.

Here my results: 

```
buffer size:    2B bytes - packet loss: 3.748444% total captured packets: 1650000
buffer size:    4B bytes - packet loss: 3.748411% total captured packets: 1650000
buffer size:    8B bytes - packet loss: 3.748444% total captured packets: 1650000
buffer size:   16B bytes - packet loss: 3.748388% total captured packets: 1650000
buffer size:   32B bytes - packet loss: 3.748500% total captured packets: 1650000
buffer size:   64B bytes - packet loss: 3.748320% total captured packets: 1650000
buffer size:  128B bytes - packet loss: 3.748469% total captured packets: 1650000
buffer size:  256B bytes - packet loss: 3.748376% total captured packets: 1650000
buffer size:  512B bytes - packet loss: 3.748388% total captured packets: 1650000
buffer size:    1k bytes - packet loss: 3.748772% total captured packets: 1650004
buffer size:    2k bytes - packet loss: 3.748444% total captured packets: 1650000
buffer size:    4k bytes - packet loss: 3.748411% total captured packets: 1650000
buffer size:    8k bytes - packet loss: 3.748346% total captured packets: 1650000
buffer size:   16k bytes - packet loss: 3.748409% total captured packets: 1650000
buffer size:   32k bytes - packet loss: 3.748409% total captured packets: 1650000
buffer size:   64k bytes - packet loss: 3.748444% total captured packets: 1650000
buffer size:  128k bytes - packet loss: 3.748388% total captured packets: 1650000
buffer size:  256k bytes - packet loss: 3.748444% total captured packets: 1650000
buffer size:  512k bytes - packet loss: 0.000000% total captured packets: 1650000
buffer size:    1M bytes - packet loss: 0.000000% total captured packets: 1650000
buffer size:    2M bytes - packet loss: 0.000000% total captured packets: 1650000
buffer size:    4M bytes - packet loss: 0.000000% total captured packets: 1650000
buffer size:    8M bytes - packet loss: 0.000000% total captured packets: 1650000
buffer size:   16M bytes - packet loss: 0.001879% total captured packets: 1650000
buffer size:   32M bytes - packet loss: 0.007575% total captured packets: 1650000
buffer size:   64M bytes - packet loss: 0.019813% total captured packets: 1650000
buffer size:  128M bytes - packet loss: 0.045427% total captured packets: 1650165
buffer size:  256M bytes - packet loss: 0.093394% total captured packets: 1650495
buffer size:  512M bytes - packet loss: 0.195770% total captured packets: 1651155
buffer size: 1024M bytes - packet loss: 0.395369% total captured packets: 1652805
```

About the hardware:
* Intel(R) Core(TM) i5-7300U CPU @ 2.60GHz
* Intel i350 network card with igb driver
* 8GB RAM

**Comment by mm120** [comment]
Hi.
I'm a colleague of sla89, and did some investigation of this issue. I'm at home now, so don't have logs and diffs, but can add them next week if needed.
My hardware and test setup was quite different, a P1010 (power pc), 512MB ram, built in ethernet hardware (gianfar). I used a much lower source of packets (flood ping) and reduced the loop in the test program to 10 (from 10000).
I got very similar results. There were three distinct phases.
1) we get a constant number of dropped packets. If I do more work in the loop, the number dropped increases.
2) things work well.
3) we seem to drop more packets as the buffer gets larger.

I added tracing to af_packet.c in the kernel. In phase 1 the packets are dropped in tpacket_rcv, which is what we expect. In this phase we have only one frame, so when the kernel pushes this frame to user space, it drops packets. The packets get dropped at the time that pcap_dispatch calls our handler.

In phase 3 it is very different. I get packets dropped in packet_rcv (no t at the start). The packets are all dropped at the start of the capture. I added more tracing and it seems to be that these packets are dropped while the kernel is allocating the memory for the capture buffer.  I assume that the flow is this:
1) open "normal" socket
2) do ioctl to set Rx ring
2.1) kernel allocates memory
2.2) kernel switches Rx function to tpacket_rcv.
I assume that the dropped packets are all those received between 1 and 2.2?

What do people think of this analysis? Is there anything that can be done do help this? Do we just have to "reset" the dropped packets counters after the pcap socket is opened?

Regards,
Mark

Ps. In phase 2, as the memory buffer increases I can see the the amount of time spent allocating the buffer does increase fairly linearly with the buffer size.

**Comment by guyharris** [comment]
> What do people think of this analysis?

Sounds correct.

> Is there anything that can be done do help this? Do we just have to "reset" the dropped packets counters after the pcap socket is opened?

That's probably the best answer; the counters should be reset as soon as ring buffer setup is complete.

**Comment by Oppen** [comment]
@mm120 your analysis seems solid to me.
What I read from is is that, while searching for pages to assign, we get packets outside of PACKET_RING mode, and we fail to read them.
Shouldn't we be able to do something like a non-blocking read from the first time, or do those packets become invalidated when we install the map?
It may seem unnecessarily complicated, but I'm worried resetting the stats may make libpcap report fewer received packets than it actually sends the user, and some programs may relay on this being consistent.

**Comment by guyharris** [comment]
> What I read from is is that, while searching for pages to assign, we get packets outside of PACKET_RING mode, and we fail to read them.
> Shouldn't we be able to do something like a non-blocking read from the first time, or do those packets become invalidated when we install the map?

Those packets are discarded when the ring is set up in the PACKET_RX_RING `setsockopt()` call.  If we wanted to process them, we'd have to read them before that call - but if packets are coming in fast enough that they'd fill up the socket buffer during the time when the ring is being set up, they're coming in fast enough that more would probably arrive after we read them all, and there's no mechanism provided by the Linux kernel that would let us atomically read the packets and set up the ring, to close that window, and it also doesn't have a mechanism to copy the packets from the socket buffer to the ring to close that window.

So we *can't* guarantee that we read all the packets that arrive outside of PACKET_RING mode.  We should probably just say "the capture starts when the ring is set up, anything before that is missed".

Unfortunately, there's no way to arrange that the statistics be cleared *and* the capture switched to "ring mode" atomically - the kernel doesn't handle that - so we can't do a perfect job of getting the statistics.

However, programs shouldn't depend on the statistics being perfect, as per...

> It may seem unnecessarily complicated, but I'm worried resetting the stats may make libpcap report fewer received packets than it actually sends the user, and some programs may relay on this being consistent.

Programs that rely on *anything* from `pcap_stats()` for any purpose other than printing out percentages are making a mistake.  To quote the man page:

       The statistics do not behave the same way on  all  platforms.   ps_recv
       might  count  packets whether they passed any filter set with pcap_set-
       filter(3PCAP) or not, or it might count only packets that pass the fil-
       ter.   It also might, or might not, count packets dropped because there
       was no room  in  the  operating  system's  buffer  when  they  arrived.
       ps_drop  is  not  available  on  all platforms; it is zero on platforms
       where it's not available.  If packet  filtering  is  done  in  libpcap,
       rather  than in the operating system, it would count packets that don't
       pass the filter.  Both ps_recv and ps_drop might, or might  not,  count
       packets not yet read from the operating system and thus not yet seen by
       the application.  ps_ifdrop might, or might  not,  be  implemented;  if
       it's  zero,  that might mean that no packets were dropped by the inter-
       face, or it might mean that the statistic is unavailable, so it  should
       not  be  treated  as  an indication that the interface did not drop any
       packets.

I put that in there to warn people not to depend too much on those statistics; all of what I said there reflects real behavior by various OSes (mostly Linux and OSes with BPF devices; I think the OSes with BPF devices all behave the same, as their BPF driver code is based on the original code written at LBL).  See the comments I added to `pcap_stats_linux()` in pcap-linux.c and `pcap_stats_bpf()` in pcap-bpf.c for details.

`pcap_stats()` needs to be deprecated in favor of a new API that returns information in a form based on [the pcapng Interface Statistics Block](http://xml2rfc.tools.ietf.org/cgi-bin/xml2rfc.cgi?url=https://raw.githubusercontent.com/pcapng/pcapng/master/draft-tuexen-opsawg-pcapng.xml&modeAsFormat=html/ascii&type=ascii#rfc.section.4.6).  That would:

- allow a platform to report only the statistics it can provide, so the caller knows what information *isn't* available;
- allow "packets that were handed to the filter" and "packets that passed the filter" to be reported separately (`isb_ifrecv` and `isb_filteraccept`), and to report both of them separately from "packets that were handed to the program using libpcap (`isb_usrdeliv`).

It would also be a Good Idea for BPF driver implementations, PF_PACKET sockets, and NPF (the WinPcap/Npcap drivers) to support reporting both "packets that were handed to the filter" and "packets that passed the filter", so that a program could report one or more of:

- "percentage of packets that passed the filter" (`isb_filteraccept*100/isb_ifrecv`)/"percentage of packets that were dropped by the filter" (`(isb_ifrecv - isb_filteraccept)*100/isb_ifrecv`);
- "percentage of packets that were dropped by the OS because we ran out of buffers" (`isb_osdrop*100/isb_ifrecv`);
- "percentage that were dropped by the adapter because it ran out of buffers" (`isb_ifdrop*100/(isb_ifdrop+isb_ifrecv)`).

("Dropped by the OS" should be more precisely specified as meaning "dropped by the OS's networking stack and capture mechanism *above the driver*; if a packet is dropped because the driver's/device's receive ring was full, that should be "dropped by the interface due to lack of resources".)

**Comment by guyharris** [comment]
Note also that PF_PACKET sockets start receiving as soon as they're created, before you've bound it to a device if you're going to be capturing on a device, so the statistics you get if you're not capturing on the "any" device may include packets received on interfaces other than the one on which you're capturing..

We do *discard* those packets - see the comment in `packet_read_packet()` in pcap-linux.c:

        /*
         * Unfortunately, there is a window between socket() and
         * bind() where the kernel may queue packets from any
         * interface.  If we're bound to a particular interface,
         * discard packets not from that interface.
         *
         * (If socket filters are supported, we could do the
         * same thing we do when changing the filter; however,
         * that won't handle packet sockets without socket
         * filter support, and it's a bit more complicated.
         * It would save some instructions per packet, however.)
         */

but we need to purge statistics for those packets as well.

(Did I mention that applications shouldn't use the counts for any reason other than calculating percentages?)

**Comment by mm120** [comment]
An option I'm thinking about is to open the socket bound to a specific protocol and then set the correct protocol only after the ring has been setup. I'm not sure if that will work cleanly, but I'll probably give it a go on Monday?
The default is ETH_P_ALL, which is 3. In other applications where I've wanted a TX only socket, I've used 0? There isn't an ETH_P_NONE?

**Comment by Oppen** [comment]
That sounds like a good idea. Port 0 is officially reserved, so we kind of can trust we won't be receiving packets targeted to that port.

**Comment by Oppen** [comment]
What if we're receiving packets for that protocol?
Maybe we can query which protocols the interface support and use one it doesn't?

**Comment by guyharris** [comment]
> An option I'm thinking about is to open the socket bound to a specific protocol and then set the correct protocol only after the ring has been setup. I'm not sure if that will work cleanly, but I'll probably give it a go on Monday?

I.e., create the socket with a protocol argument such as 0, and don't call `iface_bind()` until we've tried to set up the ring, either just before the `set_poll_timeout()` call, or after the "We set up the socket, but not with memory-mapped access." code but before the "Allocate the buffer" code, in `pcap_activate_linux()`?

When capturing on the "any" device, we *don't* bind the socket to a particular interface. [The packet(7) man page](http://man7.org/linux/man-pages/man7/packet.7.html) says, in the description of the `sockaddr_ll` structure:

       *  sll_ifindex is the interface index of the interface (see
          netdevice(7)); 0 matches any interface (only permitted for bind‐
          ing).

so setting `sll_ifindex` to 0  for the "any" device should work.

> The default is ETH_P_ALL, which is 3. In other applications where I've wanted a TX only socket, I've used 0? There isn't an ETH_P_NONE?

There isn't an ETH_P_NONE - but looking at net/packet/af_packet.c, they seem to use 0 as the protocol when they want to cause no packets to be delivered to the socket (see the ` po->num = 0;` statements).

**Comment by Oppen** [comment]
So, do we all agree using `0` as protocol at the beginning is the right choice?
If so, do you want me to implement it (as well as confirming it works as expected)?
I'll be a bit busy next two weeks, so I'll probably won't work on this (unless I get bored, which is rather likely), but if nobody has the time before that, I'm willing to.
Should we discuss it with somebody else?

**Comment by guyharris** [comment]
From reading the kernel code, it looks as if it should work.  Go ahead and try it, if either @mm120 or I haven't already done so.

**Comment by Oppen** [comment]
@guyharris, should this be closed now? I can't reproduce the issue anymore.

**Comment by guyharris** [comment]
I.e., with the current master branch libpcap, which has #887 merged, you're no longer seeing this issue?

**Comment by Oppen** [comment]
Yes, exactly.

**Comment by Oppen** [comment]
I think [this commit](https://github.com/torvalds/linux/commit/75c65772c3d18447d62d3aca5f91b06c16cc25e4) on the kernel may break it again :(

**Comment by guyharris** [comment]
> I think [this commit](https://github.com/torvalds/linux/commit/75c65772c3d18447d62d3aca5f91b06c16cc25e4) on the kernel may break it again :(

At least as of the 5.3 kernel, and perhaps before that, a change similar to that appears to be in the kernel, except that `packet_parse_headers()` *also* does that if the socket's protocol is `ETH_P_ALL`.

That code is only used on the *transmit* side, so that if you haven't specified a specific protocol, and are sending a packet, the sockbuf's protocol field is set based on the link-layer header in the raw packet data.

That code is not without some problems - see, for example, the-tcpdump-group/libpcap#739, which also notes in [one of my comments](https://github.com/the-tcpdump-group/libpcap/issues/739#issuecomment-575360259) why that pull request was submitted, and notes several fixes to various forms of breakage, and notes that there's *still* some breakage - but the breakage only seems to occur when sniffing on an interface on which another process is using a PF_PACKET socket, with a protocol of 0, to send raw link-layer packets.

So that shouldn't be a problem for this fix.
